drivers/dma/dma_mcux_edma.c

```diff
@@ -127,15 +163,20 @@ static void channel_irq(edma_handle_t *handle)
 
 		handle->header = (int8_t)new_header;
 		handle->tcdUsed -= (int8_t)tcds_done;
+
+		/* Clear the done flag before the callback so a user can optionally
+		 * install (reload) the next handle during the callback if desired
+		 */
+		if (transfer_done) {
```

Comment:
I'm not an edma expert, but I'm wondering if there's more transfer descriptors to do in the tcd, will marking the current transfer as done like this cause the hardware to move to the next one before the callback is done? That seems to go against the prior intent here and potentially problematic.

Reply:
This change does break my code, and DONE bit should not be cleared before the callback.  I am updating an I2S driver to use scatter/gather mode of eDMA, and also added dma_reload() to the eDMA driver (I am working now to merge in your updates before submitting my PR).  channel_irq() in this file is based on EDMA_HandleIRQ() from SDK eDMA driver, and we should be careful about diverging from the SDK driver.  The SDK driver clears the DONE bit after the callback for a specific reason.

When using scatter/gather mode, EDMA_SubmitTransfer() has some logic for the coherency check when TCD registers are updated.  That logic depends on the DONE bit set in some cases.  The I2S test case I am working on is one of these cases, and the SubmitTransfer() does not work properly with this update where the DONE bit is cleared before the callback.

Reply:
This change does break my code, and DONE bit should not be cleared before the callback. I am updating an I2S driver to use scatter/gather mode of eDMA, and also added dma_reload() to the eDMA driver (I am working now to merge in your updates before submitting my PR). channel_irq() in this file is based on EDMA_HandleIRQ() from SDK eDMA driver, and we should be careful about diverging from the SDK driver. The SDK driver clears the DONE bit after the callback for a specific reason.

When using scatter/gather mode, EDMA_SubmitTransfer() has some logic for the coherency check when TCD registers are updated. That logic depends on the DONE bit set in some cases. The I2S test case I am working on is one of these cases, and the SubmitTransfer() does not work properly with this update where the DONE bit is cleared before the callback.

Reply:
In the case of the UART ASYNC driver, the callback will trigger a call to DMA reload when the currently used buffer is full, and the callback requests from the user, via: [UART_RX_BUF_REQUEST](https://docs.zephyrproject.org/latest/reference/peripherals/uart.html#c.uart_event_type.UART_RX_BUF_REQUEST)

If dma_reload is called when the done bit is set (as a result of `uart_rx_buf_rsp`) then the DMA transfer is not restarted and the ping-pong transfer stops. This problem was mitigated by clearing the DONE bit before calling into the user callback.

I'll make the change back now to clear the done bit where as done in the SDK driver, but will have to look at how to correctly modify the UART driver so that transfers continue in this case.